gvfs-helper.c

```diff
@@ -632,26 +633,70 @@ static int option_parse_cache_server_mode(const struct option *opt,
 }
 
 /*
- * Let command line args override "gvfs.sharedcache" config setting.
+ * Let command line args override "gvfs.sharedcache" config setting
+ * and override the value set by git_default_config().
+ *
+ * The command line is parsed *AFTER* the config is loaded, so
+ * prepared_alt_odb() has already been called any default or inherited
+ * shared-cache has already been set.
  *
- * It would be nice to move this to parse-options.c as an
- * OPTION_PATHNAME handler.  And maybe have flags for exists()
- * and is_directory().
+ * We have a chance to override it here.
  */
 static int option_parse_shared_cache_directory(const struct option *opt,
 					       const char *arg, int unset)
 {
+	struct strbuf buf_arg = STRBUF_INIT;
+
 	if (unset) /* should not happen */
 		return error(_("missing value for switch '%s'"),
 			     opt->long_name);
 
-	if (!is_directory(arg))
-		return error(_("value for switch '%s' is not a directory: '%s'"),
-			     opt->long_name, arg);
+	strbuf_addstr(&buf_arg, arg);
+	if (strbuf_normalize_path(&buf_arg) < 0) {
+		/*
+		 * Pretend command line wasn't given.  Use whatever
+		 * settings we already have from the config.
+		 */
+		strbuf_release(&buf_arg);
+		return 0;
+	}
+	strbuf_trim_trailing_dir_sep(&buf_arg);
 
-	gvfs_shared_cache_pathname = arg;
+	if (gvfs_shared_cache_pathname && *gvfs_shared_cache_pathname &&
+	    !strcmp(buf_arg.buf, gvfs_shared_cache_pathname)) {
+		/*
+		 * The command line argument matches what we got from
+		 * the config, so we're already setup correctly. (And
+		 * we have already verified that the directory exists
+		 * on disk.)
+		 */
+		strbuf_release(&buf_arg);
+		return 0;
+	}
 
-	return 0;
+	else {
+		/*
+		 * The requested shared-cache is either different from the
+		 * one we inherited or we didn't inherit one.
+		 *
+		 * Try using/creating this shared-cache.
+		 */
+		const char *old_shared_cache = gvfs_shared_cache_pathname;
+		const char *new_shared_cache = strbuf_detach(&buf_arg, NULL);
+
+		gvfs_shared_cache_pathname = new_shared_cache;
+		add_to_alternates_memory(new_shared_cache);
+
+		/*
+		 * alt_odb_usable() nulls this if it cannot create the
+		 * directory on disk, so fallback to the previous choice
+		 * if this one failed.
+		 */
```

Comment:
I'm having some trouble following how the code manages the lifetime of `gvfs_shared_cache_pathname`.

It looks like `alt_odb_usable` will not free `gvfs_shared_cache_pathname` before setting it to NULL.  Should it be freed first? (I'm wondering if we might be leaking the memory returned by `strbuf_detach`)

Reply:
yeah, i was ok with leaking it.  i originally created the variable as a const (there were times when it was just set to something from argv during early iterations, rather than privately copying it).  And i wasn't sure it was worth the bother to refactor that.

I'll take another look and see if i can tidy it up a bit.

Reply:
I converted the global var to a strbuf which better lets me manage the underlying string (and simplified a few things too).

We will leak the final value of the string at exit() but lots of things do that.